File: src/main/java/com/itszuvalex/femtocraft/power/CryogenActiveHandler.java
package com.itszuvalex.femtocraft.power;

import com.itszuvalex.femtocraft.api.power.ICryogenHandler;
import net.minecraft.block.Block;
import net.minecraft.init.Blocks;
import net.minecraft.world.World;
import net.minecraftforge.common.util.ForgeDirection;

/**
 * Created by Chris on 9/8/2014.
 */
public class CryogenActiveHandler implements ICryogenHandler {
    public static float waterToIcePower = 100f;
    public static float lavaToObsidianPower = 300f;
    public static float airToSnowLayerPower = 10f;

    public CryogenActiveHandler() {}

    @Override
    public boolean canHandleBlock(World world, int x, int y, int z) {
        if (world.isAirBlock(x, y, z)) {
            return world.getBlock(x, y - 1, z).isSideSolid(world, x, y - 1, z, ForgeDirection.UP);
        } else {
            Block block = world.getBlock(x, y, z);
            return block == Blocks.water || block == Blocks.lava;
        }
    }

    @Override
    public float powerForBlock(World world, int x, int y, int z) {
        if (world.isAirBlock(x, y, z)) {
            if (world.getBlock(x, y - 1, z).isSideSolid(world, x, y - 1, z, ForgeDirection.UP)) {
                return airToSnowLayerPower;
            }
        } else {
            Block block = world.getBlock(x, y, z);
            if (block == Blocks.water) {
                return waterToIcePower;
            } else if (block == Blocks.lava) {
                return lavaToObsidianPower;
            }
        }
        return 0;
    }

    @Override
    public void usedBlockForPower(World world, int x, int y, int z) {
        if (world.isAirBlock(x, y, z)) {
            if (world.getBlock(x, y - 1, z).isSideSolid(world, x, y - 1, z, ForgeDirection.UP)) {
                world.setBlock(x, y, z, Blocks.snow_layer);
            }
        } else {
            Block block = world.getBlock(x, y, z);
            if (block == Blocks.water) {
                world.setBlock(x, y, z, Blocks.ice);
            } else if (block == Blocks.lava) {
                world.setBlock(x, y, z, Blocks.obsidian);
            }
        }
    }
}

File: src/main/java/com/itszuvalex/femtocraft/power/CryogenActiveHandler.scala
package com.itszuvalex.femtocraft.power

import com.itszuvalex.femtocraft.api.power.ICryogenHandler
import com.itszuvalex.femtocraft.power.CryogenActiveHandler._
import net.minecraft.init.Blocks._
import net.minecraft.world.World
import net.minecraftforge.common.util.ForgeDirection._

/**
 * Created by Chris on 9/8/2014.
 */
object CryogenActiveHandler {
  var waterToIcePower     = 100f
  var lavaToObsidianPower = 300f
  var airToSnowLayerPower = 10f
}

class CryogenActiveHandler extends ICryogenHandler {

  def canHandleBlock(world: World, x: Int, y: Int, z: Int): Boolean = {
    if (world.isAirBlock(x, y, z)) {
      world.getBlock(x, y - 1, z).isSideSolid(world, x, y - 1, z, UP)
    }
    else {
      val block = world.getBlock(x, y, z)
      block == water || block == lava
    }
  }

  def powerForBlock(world: World, x: Int, y: Int, z: Int): Float = {
    if (world.isAirBlock(x, y, z)) {
      if (world.getBlock(x, y - 1, z).isSideSolid(world, x, y - 1, z, UP)) {
        return airToSnowLayerPower
      }
    }
    else {
      val block = world.getBlock(x, y, z)
      if (block == water) {
        return waterToIcePower
      }
      else if (block == lava) {
        return lavaToObsidianPower
      }
    }
    0
  }

  def usedBlockForPower(world: World, x: Int, y: Int, z: Int) {
    if (world.isAirBlock(x, y, z)) {
      if (world.getBlock(x, y - 1, z).isSideSolid(world, x, y - 1, z, UP)) {
        world.setBlock(x, y, z, snow_layer)
      }
    }
    else {
      val block = world.getBlock(x, y, z)
      if (block == water) {
        world.setBlock(x, y, z, ice)
      }
      else if (block == lava) {
        world.setBlock(x, y, z, obsidian)
      }
    }
  }
}
